src/matlab2tikz.m

```diff
@@ -4817,22 +4830,50 @@ case size(cData,2) % specific y-data is given
             position(i,:) = axesPos ./ [figureSize, figureSize];
 
         end
-
-        % Change size if DataAspectRatioMode is manual
-        if isequal(lower(get(axesHandle,'DataAspectRatioMode')),'manual')
-            % get limits
-            xLim = get(axesHandle, 'XLim');
-            yLim = get(axesHandle, 'YLim');
-            % Get Aspect Ratio between width and height
-            aspectRatio = get(axesHandle,'DataAspectRatio');
-            % And Adjust it to the figure dimensions
-            aspectRatio = aspectRatio(1) * figWidth * (yLim(2) - yLim(1)) ...
-                / (aspectRatio(2) * figHeight * (xLim(2)-xLim(1)));
-            % Recompute height
-            newHeight = position(i,3) * aspectRatio;
-            % shrink width if newHeight is too large
-            if newHeight > position(i,4)
-                % Recompute width
+        
+        if ~(strcmpi(get(axesHandle, 'DataAspectRatioMode'), 'auto') ...
+                && strcmpi(get(axesHandle, 'PlotBoxAspectRatioMode'), 'auto'))
+            limits = axis(axesHandle);
+            if ~strcmpi(get(axesHandle, 'DataAspectRatioMode'), 'manual')
+                limits = mod(0:length(limits)-1,2); % plot box limitting mode
+                aspectRatio = get(axesHandle, 'PlotBoxAspectRatio');
+            else
+                aspectRatio = get(axesHandle, 'DataAspectRatio');
+            end
+            
+            if length(limits) == 4
+                % 2d view
+                aspectRatio = aspectRatio(1) * figWidth * (limits(4) - limits(3)) ...
+                    / (aspectRatio(2) * figHeight * (limits(2) - limits(1)));
+            else
+                % 3d view
+                
+                % project vertices of 3d plot box (this results in 2d coordinates in
+                % an absolute coordinate system that is scaled proportionally by
+                % Matlab to fit the axes position box)
+                projection = view(axesHandle);
+                vertex1 = projection * [0;0;0;1];
+                vertex2 = projection * [1;0;0;1];
+                vertex3 = projection * [0;1;0;1];
+                vertex4 = projection * [0;0;1;1];
+                vertex5 = projection * [1;1;0;1];
+                vertex6 = projection * [1;0;1;1];
+                vertex7 = projection * [0;1;1;1];
+                vertex8 = projection * [1;1;1;1];
```

Comment:
Can't we name these vertices more clearly (than just a number) or do away with those indices using some matrix calculations, e.g. as below (haven't double checked everything, though):

```matlab
projection = view(axesHandle);
vertices = projection * [0 1 0 0 1 1 0 1
                         0 0 1 0 1 0 1 1
                         0 0 0 1 0 1 1 1 
                         1 1 1 1 1 1 1 1];
verticesXY = vertices([1 2], :);

% the size of the projected plot box is limited by the long diagonals
A = [ 0  0  0 -1 +1  0  0  0
      0  0 -1  0  0 +1  0  0
      0 -1  0  0  0  0 +1  0
     -1  0  0  0  0  0  0 +1];
diagonals = verticesXY * A';
dimensions = max(abs(diagonals), [], 2);
...
```

I've also taken the liberty to re-order the diagonals (`diag1` moved behind the others, to get this nice structure).